TypeScript issue with 'third-party-capital' and google-analytics (despite not using them)

I am trying to use Nuxt Scripts for Microsoft Clarity.

Nuxt 3.12.3
Nuxt Scripts 0.6.2
TypeScript 5.5.3
vue-tsc 2.0.26

My nuxt config:
```ts
$production: {
  scripts: {
    registry: {
      clarity: true,
    },
  },
},

runtimeConfig: {
  githubToken: "",
  public: {
    scripts: {
      clarity: {
        id: "",
      },
    },
  },
},
// ...
typescript: {
  typeCheck: true,
  strict: true,
  tsConfig: {
    compilerOptions: {
      baseUrl: ".",
    },
  },
},
// ...
```

Starting dev server:
```
/Users/dinokupinic/.nvm/versions/node/v20.10.0/bin/pnpm run dev

> dinokupinic@ dev /Users/dinokupinic/Programming/TypeScript/dinokupinic
> nuxt dev

Nuxt 3.12.3 with Nitro 2.9.7                                                                                9:11:37 PM
                                                                                                                                                                         9:11:38 PM
  ➜ Local:    http://localhost:3000/
  ➜ Network:  use --host to expose

ℹ Using Tailwind CSS from ~/assets/css/tailwind.css                                                                                                    nuxt:tailwindcss 9:11:39 PM
  ➜ DevTools: press Shift + Option + D in the browser (v1.3.9)                                                                                                           9:11:42 PM

ℹ Tailwind Viewer: http://localhost:3000/_tailwind/                                                                                                    nuxt:tailwindcss 9:11:42 PM
✔ Nuxt Nitro server built in 2458 ms                                                                                                                              nitro 9:11:47 PM
ℹ Vite server warmed up in 6258ms                                                                                                                                       9:11:50 PM
ℹ ✨ new dependencies optimized: date-fns                                                                                                                               9:11:50 PM
ℹ ✨ optimized dependencies changed. reloading                                                                                                                          9:11:50 PM
ℹ Vite client warmed up in 6852ms                                                                                                                                       9:11:50 PM

 ERROR                                                                                                                                                                   9:12:00 PM
 ERROR(vue-tsc)  Cannot find module 'third-party-capital' or its corresponding type declarations.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:4:41

    2 | import { useRegistryScript } from '#nuxt-scripts-utils'
    3 | import type { RegistryScriptInput } from '#nuxt-scripts'
  > 4 | import type { GoogleAnalyticsApi } from "third-party-capital";
      |                                         ^^^^^^^^^^^^^^^^^^^^^
    5 | import { object, any } from "#nuxt-scripts-validator";
    6 | const OptionSchema = object({id:  any()})
    7 |

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:22:38

    20 |         scriptOptions: {
    21 |             use: () => {
  > 22 |           return { dataLayer: window.dataLayer, gtag: window.gtag };
       |                                      ^^^^^^^^^
    23 |         },
    24 |             stub: import.meta.client ? undefined :  ({ fn }) => {
    25 |           return fn === "dataLayer" ? [] : void 0;

 ERROR(vue-tsc)  Property 'gtag' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:22:62

    20 |         scriptOptions: {
    21 |             use: () => {
  > 22 |           return { dataLayer: window.dataLayer, gtag: window.gtag };
       |                                                              ^^^^
    23 |         },
    24 |             stub: import.meta.client ? undefined :  ({ fn }) => {
    25 |           return fn === "dataLayer" ? [] : void 0;

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:29:68

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.gtag=function gtag(){window.dataLayer.push(arguments);};window.gtag('js',new Date());window.gtag('config',options.id)},
       |                                                                    ^^^^^^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:29:85

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.gtag=function gtag(){window.dataLayer.push(arguments);};window.gtag('js',new Date());window.gtag('config',options.id)},
       |                                                                                     ^^^^^^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'gtag' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:29:106

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.gtag=function gtag(){window.dataLayer.push(arguments);};window.gtag('js',new Date());window.gtag('config',options.id)},
       |                                                                                                          ^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:29:134

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.gtag=function gtag(){window.dataLayer.push(arguments);};window.gtag('js',new Date());window.gtag('config',options.id)},
       |                                                                                                                                      ^^^^^^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'gtag' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:29:169

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.gtag=function gtag(){window.dataLayer.push(arguments);};window.gtag('js',new Date());window.gtag('config',options.id)},
       |                                                                                                                                                                         ^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'gtag' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-analytics.ts:29:198

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.gtag=function gtag(){window.dataLayer.push(arguments);};window.gtag('js',new Date());window.gtag('config',options.id)},
       |                                                                                                                                                                                                      ^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Cannot find module 'third-party-capital' or its corresponding type declarations.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-tag-manager.ts:4:42

    2 | import { useRegistryScript } from '#nuxt-scripts-utils'
    3 | import type { RegistryScriptInput } from '#nuxt-scripts'
  > 4 | import type { GoogleTagManagerApi } from "third-party-capital";
      |                                          ^^^^^^^^^^^^^^^^^^^^^
    5 | import { object, any } from "#nuxt-scripts-validator";
    6 | const OptionSchema = object({id:  any()})
    7 |

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-tag-manager.ts:22:38

    20 |         scriptOptions: {
    21 |             use: () => {
  > 22 |           return { dataLayer: window.dataLayer, google_tag_manager: window.google_tag_manager };
       |                                      ^^^^^^^^^
    23 |         },
    24 |             stub: import.meta.client ? undefined :  ({ fn }) => {
    25 |           return fn === "dataLayer" ? [] : void 0;

 ERROR(vue-tsc)  Property 'google_tag_manager' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-tag-manager.ts:22:76

    20 |         scriptOptions: {
    21 |             use: () => {
  > 22 |           return { dataLayer: window.dataLayer, google_tag_manager: window.google_tag_manager };
       |                                                                            ^^^^^^^^^^^^^^^^^^
    23 |         },
    24 |             stub: import.meta.client ? undefined :  ({ fn }) => {
    25 |           return fn === "dataLayer" ? [] : void 0;

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-tag-manager.ts:29:68

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.dataLayer.push({'gtm.start':new Date().getTime(),event:'gtm.js'});},
       |                                                                    ^^^^^^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-tag-manager.ts:29:85

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.dataLayer.push({'gtm.start':new Date().getTime(),event:'gtm.js'});},
       |                                                                                     ^^^^^^^^^
    30 |     }), _options)
    31 | }

 ERROR(vue-tsc)  Property 'dataLayer' does not exist on type 'Window & typeof globalThis'.
 FILE  /Users/dinokupinic/Programming/TypeScript/dinokupinic/.nuxt/nuxt-scripts/tpc/google-tag-manager.ts:29:106

    27 |             ...({"tagPriority":1})
    28 |         },
  > 29 |         clientInit: import.meta.server ? undefined : () => {window.dataLayer=window.dataLayer||[];window.dataLayer.push({'gtm.start':new Date().getTime(),event:'gtm.js'});},
       |                                                                                                          ^^^^^^^^^
    30 |     }), _options)
    31 | }

[vue-tsc] Found 15 errors. Watching for file changes.
```

is there a fix for this?

--- a/src/tpc/utils.ts
+++ b/src/tpc/utils.ts
@@ -2,6 +2,7 @@ import type { ExternalScript, Output, Script } from 'third-party-capital'
 import { genImport, genTypeImport } from 'knitwork'
 import { useNuxt } from '@nuxt/kit'
 import type { HeadEntryOptions } from '@unhead/vue'
+import { resolvePath } from 'mlly'
 
 export interface ScriptContentOpts {
   data: Output
@@ -21,7 +22,7 @@ export interface ScriptContentOpts {
 const HEAD_VAR = '__head'
 const INJECTHEAD_CODE = `const ${HEAD_VAR} =  injectHead()`
 
-export function getTpcScriptContent(input: ScriptContentOpts) {
+export async function getTpcScriptContent(input: ScriptContentOpts) {
   const nuxt = useNuxt()
   if (!input.data.scripts)
     throw new Error('input.data has no scripts !')
@@ -46,7 +47,9 @@ export function getTpcScriptContent(input: ScriptContentOpts) {
 
   if (input.tpcTypeImport) {
     // TPC type import
-    imports.add(genTypeImport('third-party-capital', [input.tpcTypeImport]))
+    imports.add(genTypeImport(await resolvePath('third-party-capital', {
+      url: import.meta.url,
+    }), [input.tpcTypeImport]))
   }
 
   if (hasParams) {

--- a/test/unit/tpc.test.ts
+++ b/test/unit/tpc.test.ts
@@ -38,7 +38,7 @@ describe.each([
       scriptFunctionName: 'useScriptGoogleAnalytics',
       use: () => { },
       stub: () => { },
-    })).toThrowError('no main script found for google-analytics in third-party-capital')
+    })).rejects.toThrow('no main script found for google-analytics in third-party-capital')
   })
 
   describe('script content generation', () => {
@@ -68,8 +68,8 @@ describe.each([
       },
     }
 
-    it(`expect to${isDev ? '' : ' not'} add the schema to the script options`, () => {
-      const result = getTpcScriptContent(input)
+    it(`expect to${isDev ? '' : ' not'} add the schema to the script options`, async () => {
+      const result = await getTpcScriptContent(input)
       const returnStatement = getTpcScriptReturnStatement(result, 'useScriptGoogleAnalytics')
       if (!returnStatement || returnStatement.argument?.type !== TSESTree.AST_NODE_TYPES.CallExpression || (returnStatement.argument?.callee as TSESTree.Identifier).name !== 'useRegistryScript') {
         throw new Error('TPC Scripts must return a call expression of useRegistryScript')
@@ -92,8 +92,8 @@ describe.each([
       }
     })
 
-    it('expect to stringify the use and stub functions', () => {
-      const result = getTpcScriptContent(input)
+    it('expect to stringify the use and stub functions', async () => {
+      const result = await getTpcScriptContent(input)
       const returnStatement = getTpcScriptReturnStatement(result, 'useScriptGoogleAnalytics')
       if (!returnStatement || returnStatement.argument?.type !== TSESTree.AST_NODE_TYPES.CallExpression || (returnStatement.argument?.callee as TSESTree.Identifier).name !== 'useRegistryScript') {
         throw new Error('TPC Scripts must return a call expression of useRegistryScript')
@@ -111,8 +111,8 @@ describe.each([
       expect(getCodeFromAst(result, stubFn)).toContain('return []')
     })
 
-    it('expect to augment window types', () => {
-      const result = getTpcScriptContent(input)
+    it('expect to augment window types', async () => {
+      const result = await getTpcScriptContent(input)
       const ast = parse(result, { loc: true, range: true })
       const augmentWindowTypes = ast.body.find((node): node is TSESTree.TSModuleDeclaration => node.type === TSESTree.AST_NODE_TYPES.TSModuleDeclaration)
       expect(augmentWindowTypes).toBeTruthy()
@@ -145,12 +145,12 @@ describe('script content generation with head positioning', () => {
   }
 
   describe('main script', () => {
-    it('main script post body position', () => {
-      const scriptOptsAst = getTpcScriptOptsASt(getTpcScriptContent(inputBase), 'useScriptGoogleAnalytics')
-      expect(getCodeFromAst(getTpcScriptContent(inputBase), scriptOptsAst)).toContain('"tagPosition":"bodyClose"')
+    it('main script post body position', async () => {
+      const scriptOptsAst = getTpcScriptOptsASt(await getTpcScriptContent(inputBase), 'useScriptGoogleAnalytics')
+      expect(getCodeFromAst(await getTpcScriptContent(inputBase), scriptOptsAst)).toContain('"tagPosition":"bodyClose"')
     })
-    it('main script pre body position', () => {
-      const scriptOptsAst = getTpcScriptOptsASt(getTpcScriptContent({
+    it('main script pre body position', async () => {
+      const scriptOptsAst = getTpcScriptOptsASt(await getTpcScriptContent({
         ...inputBase,
         data: {
           ...inputBase.data,
@@ -160,7 +160,7 @@ describe('script content generation with head positioning', () => {
           }],
         },
       }), 'useScriptGoogleAnalytics')
-      expect(getCodeFromAst(getTpcScriptContent(inputBase), scriptOptsAst)).toContain('"tagPosition":"bodyClose"')
+      expect(getCodeFromAst(await getTpcScriptContent(inputBase), scriptOptsAst)).toContain('"tagPosition":"bodyClose"')
     })
   })
 })

